### ISSUE TITLE
ModuleNotFoundError: Command `doctest` not found by nbqa

### ISSUE THREAD

**Issue by ntaylor-nanigans** [nbQA-dev/nbQA#363]
How do I use `nbqa` with `doctest`?  I tried `nbqa doctest mynb.ipynb` and got an error `ModuleNotFoundError: Command `doctest` not found by nbqa`  I can run doctest if I do `nbconvert` then `python -m doctest mynb.ipynb`

```
% nbqa --version
nbqa 0.3.3
% nbqa doctest mynb.ipynb
Traceback (most recent call last):
  File "/Users/ntaylor/src/nanml-notebook/docs/venv/bin/nbqa", line 8, in <module>
    sys.exit(main())
  File "/Users/ntaylor/src/nanml-notebook/docs/venv/lib/python3.8/site-packages/nbqa/__main__.py", line 531, in main
    output_codes = [
  File "/Users/ntaylor/src/nanml-notebook/docs/venv/lib/python3.8/site-packages/nbqa/__main__.py", line 532, in <listcomp>
    _run_on_one_root_dir(i, cli_args, configs, project_root)
  File "/Users/ntaylor/src/nanml-notebook/docs/venv/lib/python3.8/site-packages/nbqa/__main__.py", line 474, in _run_on_one_root_dir
    out, err, output_code, mutated = _run_command(
  File "/Users/ntaylor/src/nanml-notebook/docs/venv/lib/python3.8/site-packages/nbqa/__main__.py", line 355, in _run_command
    raise ModuleNotFoundError(_get_command_not_found_msg(command))
ModuleNotFoundError: Command `doctest` not found by nbqa.

Please make sure you have it installed in the same Python environment as nbqa. See
e.g. https://realpython.com/python-virtual-environments-a-primer/ for how to set up
a virtual environment in Python.

Since nbqa is installed at /Users/ntaylor/src/nanml-notebook/docs/venv/lib/python3.8/site-packages/nbqa and uses the Python executable found at
/Users/ntaylor/src/nanml-notebook/docs/venv/bin/python, you could fix this issue by running `/Users/ntaylor/src/nanml-notebook/docs/venv/bin/python -m pip install doctest`.
```

**Comment by MarcoGorelli** [comment]
Hi @ntaylor-nanigans 

Thanks for the report, I really appreciate it!

I just tried running `nbqa doctest` on a notebook I have locally and it seemed to work fine, so I don't know why it failed for you.

Were you running `nbqa` from the command line?  Are you able to share your notebook by any chance so I can try running `nbqa doctest mynb.ipynb` myself?

**Comment by MarcoGorelli** [comment]
I _think_ I have an idea, actually - does your doctest import something it's not able to find?

The problem is that to detect whether you have the code quality tool you're running installed or not, we do
```
    output = subprocess.run(
        [sys.executable, "-m", command, str(arg), *cmd_args],
        stderr=subprocess.PIPE,
        stdout=subprocess.PIPE,
        cwd=tmpdirname,
        env=env,
    )

    mutated = _get_mtimes(arg) != before

    output_code = output.returncode

    out = output.stdout.decode()
    err = output.stderr.decode()

    if "No module named" in err:
        raise ModuleNotFoundError(_get_command_not_found_msg(command))
```

However, this is buggy, because `No module named` might be returned by the code quality tool itself! So, we have a bug.

In the short-run, I'd recommend fixing your docstring so that everything you use in it was imported correctly. We'll make a patch release for this at the weekend - @s-weigand had suggested a better way of checking if packages are installed, and I wish I'd followed his advice immediately now :flushed:

@all-contributors please add @ntaylor-nanigans for bug

**Comment by girip11** [comment]
@MarcoGorelli  I think you are right. If I had printed `err` as part of the message of `ModuleNotFoundError` change that I pushed, we would have got a very good view now.

Do you have any fixes in mind? 

Quickest one I can suggest is to change the string to `No module named '{command}'`. But that's hacky again. 
Or We could use `importlib.util.find_spec("command")` to check if the module exists or not before spawning the subprocess.

**Comment by MarcoGorelli** [comment]
> Or We could use importlib.util.find_spec("command") to check if the module exists or not before spawning the subprocess.

This sounds safer - I can give a go at this tomorrow morning

**Comment by MarcoGorelli** [comment]
I can't reproduce this. I've tried making a notebook with
```python
def f():
    """
    >>> from x import y
    """
```
and I get
```python-traceback
File "tests/data/invalid_import_in_doctest.ipynb", cell_1:3, in invalid_import_in_doctest.f
Failed example:
    from x import y
Exception raised:
    Traceback (most recent call last):
      File "/usr/lib/python3.8/doctest.py", line 1336, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest invalid_import_in_doctest.f[0]>", line 1, in <module>
        from x import y
    ModuleNotFoundError: No module named 'x'
**********************************************************************
1 items had failures:
   1 of   1 in invalid_import_in_doctest.f
***Test Failed*** 1 failures.
```

So @ntaylor-nanigans any extra info to help us figure out where the bug is would be really appreciated :smile:

**Comment by ntaylor-nanigans** [comment]
@MarcoGorelli thanks for taking a look!  I can replicate your example successfully, but then when I run it against the notebook I want to work on, I get the error.  Still, if I run in on yet another Notebook I'm working on, then I don't get an error.

So there's something about the notebook—I don't want to waste your time—are there any known things that would cause trouble?  I tried removing some magic `%run` on a whim, but that didn't do the trick.

```
(venv) % nbqa doctest /tmp/test.ipynb
**********************************************************************
File "/tmp/test.ipynb", cell_1:5, in test.f
Failed example:
    from x import y
Exception raised:
    Traceback (most recent call last):
      File "/usr/local/Cellar/python@3.8/3.8.6/Frameworks/Python.framework/Versions/3.8/lib/python3.8/doctest.py", line 1336, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest test.f[0]>", line 1, in <module>
        from x import y
    ModuleNotFoundError: No module named 'x'
**********************************************************************
1 items had failures:
   1 of   1 in test.f
***Test Failed*** 1 failures.

(venv) % nbqa doctest jupyter/lib/nanml/dataset.ipynb
Traceback (most recent call last):
  File "/Users/ntaylor/src/nanml-notebook/docs/venv/bin/nbqa", line 8, in <module>
    sys.exit(main())
  File "/Users/ntaylor/src/nanml-notebook/docs/venv/lib/python3.8/site-packages/nbqa/__main__.py", line 531, in main
    output_codes = [
  File "/Users/ntaylor/src/nanml-notebook/docs/venv/lib/python3.8/site-packages/nbqa/__main__.py", line 532, in <listcomp>
    _run_on_one_root_dir(i, cli_args, configs, project_root)
  File "/Users/ntaylor/src/nanml-notebook/docs/venv/lib/python3.8/site-packages/nbqa/__main__.py", line 474, in _run_on_one_root_dir
    out, err, output_code, mutated = _run_command(
  File "/Users/ntaylor/src/nanml-notebook/docs/venv/lib/python3.8/site-packages/nbqa/__main__.py", line 355, in _run_command
    raise ModuleNotFoundError(_get_command_not_found_msg(command))
ModuleNotFoundError: Command `doctest` not found by nbqa.

Please make sure you have it installed in the same Python environment as nbqa. See
e.g. https://realpython.com/python-virtual-environments-a-primer/ for how to set up
a virtual environment in Python.

Since nbqa is installed at /Users/ntaylor/src/nanml-notebook/docs/venv/lib/python3.8/site-packages/nbqa and uses the Python executable found at
/Users/ntaylor/src/nanml-notebook/docs/venv/bin/python, you could fix this issue by running `/Users/ntaylor/src/nanml-notebook/docs/venv/bin/python -m pip install doctest`.
```

**Comment by MarcoGorelli** [comment]
> I don't want to waste your time

**Comment by MarcoGorelli** [comment]
> I don't want to waste your time

Quite the opposite, bug reports are incredibly useful - thank you!

> when I run it against the notebook I want to work on, I get the error.

**Comment by ntaylor-nanigans** [comment]
I'd prefer not to share it with the world, so here is a Drive link. Can you request access?  https://drive.google.com/file/d/19CgtL758pPkoz28Q_aW3bHTzd-JYuI8j/view

**Comment by ntaylor-nanigans** [comment]
Also, I should have noted: I am in the process of adding `doctest` and revision I sent you doesn't actually contain any `doctest` which I now realize is terrible of me, but it does seem that I need it to run without any tests before it will run with any tests.

**Comment by ntaylor-nanigans** [comment]
Double sorry, I missed all the discussion and just saw the last post.  YES - there are definitely missing imports!

**Comment by MarcoGorelli** [comment]
Awesome - thank you so much for the notebook, I'll try to come up with a minimal example which reproduces the bug and'll make a new release as soon as that's done, will keep you posted!

**Comment by MarcoGorelli** [comment]
closed in #365 

@ntaylor-nanigans please try upgrading to version 0.3.4, I'd like to think this is fixed - if there are issues, please do let us know!
